src/app/games/apagos/ApagosState.ts

```diff
@@ -0,0 +1,80 @@
+import { GameState } from 'src/app/jscaip/GameState';
+import { Player } from 'src/app/jscaip/Player';
+import { ArrayUtils, NumberTable } from 'src/app/utils/ArrayUtils';
+import { MGPMap } from 'src/app/utils/MGPMap';
+import { ApagosCoord } from './ApagosCoord';
+import { ApagosSquare } from './ApagosSquare';
+
+export class ApagosState extends GameState<ApagosCoord, ApagosSquare> {
+
+    public static PIECE_BY_PLAYER: number = 10;
+
+    public static getInitialState(): ApagosState {
+        return ApagosState.fromRepresentation(0, [
+            [0, 0, 0, 0],
+            [0, 0, 0, 0],
+            [7, 5, 3, 1],
+        ], this.PIECE_BY_PLAYER, this.PIECE_BY_PLAYER);
+    }
+    public static fromRepresentation(turn: number, board: NumberTable, nbZero: number, nbOne: number): ApagosState {
+        const squares: ApagosSquare[] = [];
+        for (let x: number = 0; x < 4; x++) {
+            const nbZero: number = board[0][x];
+            const nbOne: number = board[1][x];
+            const nbTotal: number = board[2][x];
+            const square: ApagosSquare = ApagosSquare.from(nbZero, nbOne, nbTotal).get();
+            squares.push(square);
+        }
+        const remaining: MGPMap<Player, number> = new MGPMap();
+        remaining.set(Player.ZERO, nbZero);
+        remaining.set(Player.ONE, nbOne);
+        return new ApagosState(turn, squares, remaining);
+    }
+    public constructor(turn: number,
+                       public readonly board: ReadonlyArray<ApagosSquare>,
+                       public readonly remaining: MGPMap<Player, number>)
+    {
+        super(turn);
+        this.remaining.makeImmutable();
+    }
+    public getPieceAt(coord: ApagosCoord): ApagosSquare {
+        return this.board[coord.x];
+    }
+    public updateAt(coord: ApagosCoord, newSquare: ApagosSquare): ApagosState {
+        const newBoard: ApagosSquare[] = [];
+        for (let x: number = 0; x < 4; x++) {
+            if (coord.x === x) {
+                newBoard.push(newSquare);
+            } else {
+                newBoard.push(this.board[x]);
+            }
+        }
+        const remaining: MGPMap<Player, number> = this.getRemainingCopy();
+        return new ApagosState(this.turn, newBoard, remaining);
+    }
+    public getRemainingCopy(): MGPMap<Player, number> {
+        const nbZero: number = this.getRemaining(Player.ZERO);
+        const nbOne: number = this.getRemaining(Player.ONE);
+        const remaining: MGPMap<Player, number> = new MGPMap();
+        remaining.set(Player.ZERO, nbZero);
+        remaining.set(Player.ONE, nbOne);
+        return remaining;
+    }
+    public isOnBoard(coord: ApagosCoord): boolean {
+        return coord != null;
+    }
+    public getNullable(coord: ApagosCoord): ApagosSquare {
+        if (coord == null) {
```

Comment:
est-ce que c'est reachable ça ? si oui, est-ce que c'est souhaitable d'appeler avec null ?

Reply:
Non c'est pas reachable, en fait, l'entierté de la méthode est non reachable, donc ce qui est dedans whatever, j'peux mettre un assert(true, 'useless poop'); si besoin :D

Reply:
mais comme toutes les ApagosCoord sont crée d'avance c'est un nullable qu'est ... bah jamais null du coup, c'qui est antinomique (pas sûr de l'usage du mot)

Reply:
dans `strictness` j'ai supprimé le fait de devoir définir `getNullable` par défaut. Les jeux qui en ont besoin peuvent le définir, et les autres peuvent s'en foutre, donc on pourra la supprimer !

Reply:
Oui, elle aura eu une courte vie cette fonction :p donc là quoi, je met un TODO "delete when strictness is merged" ?

Reply:
yep peut-être pour être sur !

Reply:
Gardera-t-on isOnBoard ?

Reply:
hmm idem je dirais : s'il est pas nécessaire de l'avoir dans l'interface de la classe abstraite, non, et juste le définir dans les classes filles ? (par exemple que GameStateWithTable la définisse c'est bien, au pire on l'utilise pas mais elle est testée dans GameStateWithTable.spec.ts, mais qu'elle soit require par GameState alors qu'elle est pas nécessaire, non). À sous-ticketter dans strictness si tu es accord !

Reply:
J'ai mis des TODO qui font qu'on saura vite supprimer les non nécessaire.
Mais sinon, on peut pas aussi faire simplement une fonction qui throw si elle n'est pas overridée ?

Reply:
on pourrait mais pas trop d'intérêt non ? C'est mieux d'avoir une erreur de compil si on l'utilise sans la définir plutôt qu'un throw